README.md

```diff
@@ -1,41 +1,95 @@
-# Crate Name
+# vm-allocator
 
-## Design
+`vm-allocator` is a crate designed to to provide allocation and release strategies
+that are needed by the VMM during the lifetime of a virtual machine. Possible
+resource types that a VMM could allocate using vm-allocator are MMIO addresses,
+PIO addresses, IRQ numbers, MSI vectors, device ids.
 
-TODO: This section should have a high-level design of the crate.
+Main components are:
 
-Some questions that might help in writing this section:
-- What is the purpose of this crate?
-- What are the main components of the crate? How do they interact which each
-  other?
+- `IDAllocator` - which should be used for all resources that can be reduced to
+an integer type.
+- `AddressAllocator` - which should be used to allocate address ranges in different
+address spaces. This component is a wrapper over `IntervalTree` that adds semantic meaning
+for address ranges.
 
-## Usage
+## ID Allocator
 
-TODO: This section describes how the crate is used.
+### Design
 
-Some questions that might help in writing this section:
-- What traits do users need to implement?
-- Does the crate have any default/optional features? What is each feature
-  doing?
-- Is this crate used by other rust-vmm components? If yes, how?
+This allocator should be used to allocate resources that can be reduced to an integer
+type like legacy IRQ numbers, MSI interrupt numbers or KVM memory slot IDs. The
+characteristics of such a resource are represented by the `IdAllocator` struct.
+
+```rust
+/// Id allocator representation.
+pub struct IdAllocator {
```

Comment:
If we duplicate part of the implementation in README, then we need to make sure to keep them in sync if something changes (method signatures, doc comments, etc).

Reply:
We can also just move the examples either in the code, or in an `examples` directory. A layout example is available here: https://doc.rust-lang.org/cargo/guide/project-layout.html